aip/auth/4117.md

```diff
@@ -0,0 +1,387 @@
+
+credentials, project = google.auth.default()
+```
+
+External account credentials can also be initialized explicitly using the
+generated configuration file.
+
+```python
+# Sample for Azure or OIDC/SAML providers.
+import json
+
+from google.auth import identity_pool
+
+json_config_info = json.loads(function_to_get_json_config())
+credentials = identity_pool.Credentials.from_info(json_config_info)
+scoped_credentials = credentials.with_scopes(
+    ['https://www.googleapis.com/auth/cloud-platform'])
+```
+
+```python
+# Sample for AWS.
+import json
+
+from google.auth import aws
+
+json_config_info = json.loads(function_to_get_json_config())
+credentials = aws.Credentials.from_info(json_config_info)
+scoped_credentials = credentials.with_scopes(
+    ['https://www.googleapis.com/auth/cloud-platform'])
+```
+
+### Expected Behavior
+
+The auth libraries will use the information in the JSON configuration file to
+retrieve the external credentials and exchange them for Google access tokens
+using the GCP Security Token Service (via the token exchange endpoint
+`https://sts.googleapis.com/v1/token`) and then impersonating a service account
+by calling the **IamCredentials** [generateAccessToken][5] API to access GCP
+resources.
+
+All external account JSON files share the following fields:
+
+| Field Name                        | Required | Description |
+|-----------------------------------|----------|-------------|
+| type                              | Yes      | This identifies the new type of credential object. This is always "external_account" |
+| audience                          | Yes      | This is the STS audience which contains the resource name for the workload identity pool and the provider identifier in that pool. |
+| subject_token_type                | Yes      | This is the STS subject token type based on the [OAuth 2.0 token exchange spec][7]. |
+| service_account_impersonation_url | No       | This is the URL for the service account impersonation request. If this is not available, the STS returned access token is directly used without impersonation. |
+| token_url                         | Yes      | This is the STS token exchange endpoint. |
+| credential_source.*               | Yes      | This object defines the mechanism used to retrieve the external credential from the local environment so that it can be exchanged for a GCP access token via the STS endpoint. |
+
+The auth libraries and applications **must** follow the steps below for all
+types of external account credentials:
+
+- Check **credential_source** to determine the necessary logic to retrieve the
+  external credential which will be used to construct the subject token to pass
+  to the STS endpoint. This is covered in detail for every credential
+  configuration below.
+- Construct the STS request, based on [rfc8693][4]:
+  - STS audience is constructed using the **audience** field.
+  - **grant_type** is always `urn:ietf:params:oauth:grant-type:token-exchange`
```

Comment:
s/is always/\*\*must\*\* be/

Same comment for other occurrences.

Reply:
Done